console.go

```diff
@@ -0,0 +1,173 @@
+	"errors"
+	"fmt"
+	"net/url"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws/credentials"
+	"github.com/miquella/vaulted/lib"
+	"github.com/pkg/browser"
+)
+
+var (
+	ErrInvalidTemporaryCredentials = errors.New("Temporary session credentials found. Console cannot be opened with temporary credentials.\nIf the AWS Key is set in a vault, the permanent credentials can be used by specifying the vault name.")
+	ErrInvalidDuration             = errors.New("Console duration must be between 15m and 12h.")
+	ErrNoCredentialsFound          = errors.New("No credentials found. Console cannot be opened.")
+)
+
+const (
+	ConsoleURL                 = "https://console.aws.amazon.com/console/home"
+	ConsoleFederationSigninURL = "https://signin.aws.amazon.com/federation"
+
+	ConsoleMinDuration     = 15 * time.Minute
+	ConsoleMaxDuration     = 12 * time.Hour
+	ConsoleDefaultDuration = 1 * time.Hour
+)
+
+type Console struct {
+	VaultName string
+	Role      string
+	Duration  time.Duration
+}
+
+func (c *Console) Run(store vaulted.Store) error {
+	signinToken, err := c.getSigninToken(store)
+	if err != nil {
+		return err
+	}
+
+	return openConsole(signinToken)
+}
+
+func (c *Console) getSigninToken(store vaulted.Store) (string, error) {
+	vault, err := c.getVault(store)
+	if err != nil {
+		return "", err
+	}
+
+	awsKey := c.getAWSKey(vault)
+
+	awsCreds, err := c.getCredentials(vault, awsKey)
+	if err != nil {
+		return "", err
+	}
+
+	duration, err := c.getDuration(vault)
+	if err != nil {
+		return "", err
+	}
+
+	if c.Role != "" {
+		return c.getAssumeRoleToken(store, awsKey, awsCreds, duration)
+	} else {
+		return c.getFederationToken(awsCreds, duration)
+	}
+
+}
+
+func (c *Console) getVault(store vaulted.Store) (*vaulted.Vault, error) {
+	vault := &vaulted.Vault{}
+	var err error
+	if c.VaultName != "" {
+		vault, _, err = store.OpenVault(c.VaultName)
+		if err != nil {
+			return nil, err
+		}
+	}
+	return vault, nil
+}
+
+func (c *Console) getCredentials(vault *vaulted.Vault, awsKey *vaulted.AWSKey) (*vaulted.AWSCredentials, error) {
+	awsCreds, err := awsKey.AWSCredentials.WithLocalDefault()
+	if err != nil {
+		if err != credentials.ErrNoValidProvidersFoundInChain {
+			return nil, err
+		} else if err == credentials.ErrNoValidProvidersFoundInChain || !awsCreds.Valid() {
+			return nil, ErrNoCredentialsFound
+		}
+	} else if awsCreds.ValidSession() {
+		return nil, ErrInvalidTemporaryCredentials
+	}
+
+	return awsCreds, nil
+}
+
+func (c *Console) getAWSKey(vault *vaulted.Vault) *vaulted.AWSKey {
+	var awsKey vaulted.AWSKey
+
+	if vault.AWSKey != nil && vault.AWSKey.Valid() {
+		awsKey = *vault.AWSKey
+	}
+	awsKey.Role = c.Role
+
+	return &awsKey
+}
+
+func (c *Console) getDuration(vault *vaulted.Vault) (time.Duration, error) {
+	var duration time.Duration
+	if c.Duration != 0 {
+		duration = c.Duration
+	} else if vault.Duration != 0 {
+		duration = vault.Duration
+	} else {
+		duration = ConsoleDefaultDuration
+	}
+
+	return capDuration(duration)
```

Comment:
Even if we've validated the Duration when we received the command, it still feels like we shouldn't be calling capDuration on the command flag duration.  capDuration really only belongs with the Vault duration.